src/tomato/drivers/biologic/main.py

```diff
@@ -2,16 +2,94 @@
 import multiprocessing
 import time
 from filelock import FileLock
-
 from datetime import datetime, timezone
-
 from .kbio_wrapper import (
     get_kbio_techpath,
     payload_to_ecc,
     parse_raw_data,
     get_kbio_api,
 )
 
+def safe_api_connect(api,
+                    address: str,
+                    retries: int = 100,
+                    timeout: int = 10
```

Comment:
The default values - `retries` and `timeout` could be in principle pulled in from the device settings, the same way the `dllpath` is. I would say a `retries` count of 100 is unreasonably high as a default, but allowing user customisation via the settings file would be an easy way around it.

Reply:
I am not sure whether the dllpath is the path to the setting file? If this is the case, then I imagine I would just have to join this dllpath thie setting.yaml. (os.path.join(dllpath,'setting.yaml')) Then I will write a commad to open the setting file and retrive retires and timeout (which would be now be specified by the user there), then use it here. Is this the preferred method you are suggesting?

Reply:
See comment below.

Reply:
Ok, I have make changes as suggested. Now, the retries and time_out paraemters will be obtained from the setting file (the same way as dllpath). I also put a default value here as reduce the number of default value from 100 to 10.